src/zope/publisher/tests/test_http.py

```diff
@@ -896,6 +896,13 @@ def testContentType(self):
         eq("application/json", headers["Content-Type"])
         eq(b"test", body)
 
+        # zope.app.exception.browser.unauthorized in combination with
+        # zope.pluggableauth.plugins.session.SessionCredentialsPlugin (which
+        # redirects to login form on 401/403) produces a None body
+        response = HTTPResponse()
+        response.setResult(None)
+        eq(b"", response.consumeBody())
```

Comment:
Perhaps also add

```python
        response = HTTPResponse()
        response.setResult(b'')
        eq(b"", response.consumeBody())

        response = HTTPResponse()
        response.setResult(u'')
        eq(b"", response.consumeBody())
```

and then you'll see failures.

Now maybe setResult(u'') is not valid without specifying a Content-Type: text/something  header, but the failure should state that via this ValueError:
https://github.com/zopefoundation/zope.publisher/blob/69902a756c425bac12b73c19323c46d9ef671411/src/zope/publisher/http.py#L856-L858
and not fail with a TypeError: expected string or bytes-like object

(For a bit I was afraid that this was my bug, but no, somebody else refactored my good code and introduced this in 2010 ;)

Reply:
Or you could say that this is an unrelated bug and punt it for somebody else to fix in a separate PR?  The fix is simple:

```diff
diff --git a/src/zope/publisher/http.py b/src/zope/publisher/http.py
index 071c4b7..f763194 100644
--- a/src/zope/publisher/http.py
+++ b/src/zope/publisher/http.py
@@ -851,7 +851,7 @@ class HTTPResponse(BaseResponse):
         content_type = self.getHeader('content-type')
 
         if isinstance(body, unicode):
-            ct = content_type
+            ct = content_type or ''
             if not unicode_mimetypes_re.match(ct):
                 raise ValueError(
                     'Unicode results must have a text, RFC 3023, RFC 4627,'
```